src/box/iproto.cc

```diff
@@ -3729,6 +3758,10 @@ uint64_t
 iproto_session_new(struct iostream *io, struct user *user)
 {
 	assert(iostream_is_initialized(io));
+	if (iproto_is_shutting_down) {
+		diag_set(ClientError, ER_SHUTDOWN);
+		return -1;
```

Comment:
Changed to return sid in output argument to be able to return error as discussed f2f.

<details>
<summary>diff</summary>

```diff
@@ -3754,8 +3746,8 @@ iproto_set_msg_max(int new_iproto_msg_max)
        return 0;
 }

-uint64_t
-iproto_session_new(struct iostream *io, struct user *user)
+int
+iproto_session_new(struct iostream *io, struct user *user, uint64_t *sid)
 {
        assert(iostream_is_initialized(io));
        if (iproto_is_shutting_down) {
@@ -3773,7 +3765,8 @@ iproto_session_new(struct iostream *io, struct user *user)
        static int thread = 0;
        thread = (thread + 1) % iproto_threads_count;
        iproto_do_cfg_async(&iproto_threads[thread], cfg_msg);
-       return session->id;
+       *sid = session->id;
+       return 0;
 }

--- a/src/box/iproto.h
+++ b/src/box/iproto.h
@@ -145,8 +145,8 @@ int
 iproto_set_msg_max(int iproto_msg_max);

 /**
- * Creates a new IPROTO session over the given IO stream and returns the new
- * session id. Never fails. Doesn't yield.
+ * Creates a new IPROTO session over the given IO stream. Doesn't yield.
+ * Set the output parameter sid to the sid of newly created session.
  *
  * The IO stream must refer to a non-blocking socket but this isn't enforced by
  * this function. If it isn't so, the new connection may not work as expected.
@@ -155,16 +155,18 @@ iproto_set_msg_max(int iproto_msg_max);
  * the specified user. Otherwise, it will be authenticated as guest.
  *
  * The function takes ownership of the passed IO stream by moving it to the
- * new IPROTO connection (see iostream_move).
+ * new IPROTO connection (see iostream_move) except for error case.
  *
  * Essentially, this function passes the IO stream to the callback invoked
  * by an IPROTO thread upon accepting a new connection on a listening socket.
  * The callback creates a new IPROTO connection, attaches it to the given
  * session, then sends the greeting message and starts processing requests as
  * usual. All of this is done asynchronously by an IPROTO thread.
+ *
+ * Returns 0 on success, and -1 otherwise (diagnostic is set).
  */
-uint64_t
-iproto_session_new(struct iostream *io, struct user *user);
+int
+iproto_session_new(struct iostream *io, struct user *user, uint64_t *sid);
```

</details>